File: 852-peak-index-in-a-mountain-array.py
class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        low, high = 1, len(arr) - 2
        
        while low <= high:
            median = (low + high) >> 1
            
            if arr[median - 1] < arr[median] and arr[median + 1] < arr[median]:
                return median
            
            if arr[median - 1] > arr[median]:
                high = median - 1
            elif arr[median + 1] > arr[median]:
                low = median + 1

        return -1

File: 852-peak-index-in-a-mountain-array.go

func peakIndexInMountainArray(nums []int) int {
    low, high := 1, len(nums) - 2

    for low <= high {
        median := (low + high) >> 1

        if nums[median-1] < nums[median] && nums[median+1] < nums[median] {
            return median
        }
        if nums[median - 1] > nums[median] {
            high = median - 1
        } else if nums[median + 1] > nums[median] {
            low = median + 1
        }
    }
    return -1
}
